Refactor the spec (#24)
ShivanKaul/star-spec: draft-dss-star.md

Randomness Server:
: An entity that runs an oblivious pseudorandom function ({{!OPRF=I-D.irtf-cfrg-voprf}}) service that allows clients to receive pseudorandom function evaluations on their measurement and the server OPRF key, without the Randomness Server learning anything about their measurement. The clients use the output as randomness to produce the report that is then sent to the Aggregation Server.

Anonymizing Server:
: An entity that clients use to decouple their identity (IP address) from their messages sent to the Aggregation Server.

Client:
: The entity that provides user data to the system.

Measurement:
: The unencrypted, potentially-sensitive data that the client is asked to report.

Report:
: The encrypted measurement being sent by the client.

Auxiliary Data:
: Arbitrary data that clients may send as part of their report, but which is only revealed when at least K encrypted measurements of the same value are received.

REPORT_THRESHOLD:
: The minimum number of reports that an Aggregation Server needs before revealing client data. This value is chosen by the application.

# Cryptographic Dependencies

> STAR depends on the following cryptographic protocols and primitives:

Reviewer: This looks good to me - is it common to have this kind of section in protocol documents that deal with crypto primitives?

    can raise a `MessageLimitReachedError` upon failure.
- `Open(key, nonce, aad, ct)`: Decrypt ciphertext and tag `ct` using
  associated data `aad` with symmetric key `key` and nonce `nonce`,
  returning plaintext message `pt`. This function can raise an
  `OpenError` or `MessageLimitReachedError` upon failure.
- `Nk`: The length in bytes of a key for this algorithm.
- `Nn`: The length in bytes of a nonce for this algorithm.
- `Nt`: The length in bytes of the authentication tag for this algorithm.

[[OPEN ISSUE: specify which KCAEAD to use]]

# System Overview

In STAR, clients generate encrypted measurements and send them to a single untrusted
Aggregation Server in a report. Each report is effectively a random k-out-of-n share of
the client data secret, along with some additional auxilary data. In a given amount of
time, if the Aggregation Server receives the same encrypted value from k = REPORT_THRESHOLD
clients, the server can recover the client data associated with each report. This ensures
that clients only have their measurements revealed if they are part of a larger crowd,
thereby achieving k-anonymity privacy (where k = REPORT_THRESHOLD).

> Each client report is as secret as the underlying underlying client data. That means low

Reviewer: ```suggestion
Each client report is as secret as the underlying client data. That means low
```

entropy client data values could be abused by an untrusted Aggregation Server in a
dictionary attack to recover client data with fewer than REPORT_THRESHOLD honestly generated

> reports. To mitigate this, clients boost the entropy of their data using an Oblivious

Reviewer: ```suggestion
reports. To mitigate this, clients boost the entropy of their data using output from an Oblivious
```

Pseudorandom Function (OPRF) provided by a separate, non-colluding Randomness Server.

> STAR also requires use of a client Anonymizing Server when interacting with the Aggregation

Reviewer: ```suggestion
STAR also requires use of a client Anonymizing Proxy when interacting with the Aggregation
```

Reviewer: I see that DAP also uses this term: https://www.ietf.org/archive/id/draft-ietf-ppm-dap-01.html#name-anonymizing-proxies

~~~~ aasvg

     +------------+         +--------------+             +-------------+
     |   Client   |         |  Randomness  |             | Aggregation |
     | (msg, aux) |         |    Server    |             |   Server    |
     +---+--------+         +------+-------+             +------+------+
         |                         |                            |
         |                         |===========\                |
         | Request(Blind(msg))     |           |                |
         +------------------------>|           | Randomness     |
         |                         | Evaluate  | Phase          |
         |           Response(...) |           |                |
         |<------------------------+           |                |
         |                         |===========/                |
         |                        ...                           |
    Generate Report                                             |
    using randomness                                            |
         |                  +--------------+                    |
         |                  |  Anonymizing |                    |

>          |                  |    Server    |                    |

Reviewer: ```suggestion
         |                  |    Proxy    |                    |
```

the server computes:

```
evaluated_element = server_context.Evaluate(sk, blinded_element)
```

The Randomness Server then computes `randomness_response = SerializeElement(evaluated_element)` and
sends it to the client using the content type "message/star-randomness-response". An example
response is below.

~~~
:status = 200
content-type = message/star-randomness-response
content-length = Ne

<Bytes containing a serialized evaluated element>
~~~

Upon receipt, the client computes `evaluated_element = DeserializeElement(randomness_response)`.
If this fails, the client aborts the protocol. Otherwise, the client
finalizes the OPRF protocol to compute the output `rand` as follows:

```
rand = client_context.Finalize(msg, blind, evaluated_element)
```

## Reporting Phase {#client-message}

In the reporting phase, the client uses its measurement `msg` with auxiliary data `aux`
and its derived randomness `rand` to produce a report for the Aggregation Server.

### Reporting Configuration

The reporting phase requires the Aggregation Server to be configured with a URI for
accepting reports. As an example, the Aggregation Server URI might be https://aggregator.example.

> The Aggregation Server is both an Oblivious HTTP Target and Oblivious Gateway Resource.

Reviewer: I think we could mention something like "If we use OHTTP", but this seems like we're requiring OHTTP.

> Clients are also configured with the URI for an Anonymizing Server, which is an Oblivious

Reviewer: same here re: use of OHTTP

is only revealed when the k-condition is met but, importantly, is not part of the k-condition
itself. This data might be unique to some or all of the submissions, or omitted entirely. This
can even be the actual measured value itself. For example: if we're measuring tabs open on a
client, then the measurement being sent can be "city: Vancouver" and the aux data can be "7"
for a particular client. The idea being, that we only reveal all the measurements once we
know that there are at least K clients with city: Vancouver.

# Security Considerations {#security-considerations}

This section contains security considerations for the draft.

## Randomness Sampling {#sec-randomness-sampling}

Deterministic randomness MUST be sampled by clients to construct their STAR report, as discussed
in {{client-message}}. This randomness CANNOT be derived locally, and MUST be sampled from the
Randomness Server (that runs an {{!OPRF=I-D.irtf-cfrg-voprf}} service).

For best-possible security, the Randomness Server SHOULD sample and use a new OPRF key for each
time epoch `t`, where the length of epochs is determined by the application. The previous OPRF

> key that was used in epoch `t-1` can be safely deleted. As discussed in {{leakage}}, shorter
> epochs provide more client security, but also reduce the window in which data collection occurs.

Reviewer: ```suggestion
key that was used in epoch `t-1` can be safely deleted. As discussed in {{leakage}}, shorter
epochs provide more protection from Aggregation Server attacks, but also reduce the window in which data collection occurs and the possibility that we will have enough reports to decrypt, as well as the reporting latency.
```

: An entity that would like to learn aggregated data from users.

Randomness Server:
: An entity that runs an oblivious pseudorandom function ({{!OPRF=I-D.irtf-cfrg-voprf}})
  service that allows clients to receive pseudorandom function evaluations on their
  measurement and the server OPRF key, without the Randomness Server learning anything
  about their measurement. The clients use the output as randomness to produce the
  report that is then sent to the Aggregation Server.

Anonymizing Server:
: An entity that clients use to decouple their identity (IP address) from their
messages sent to the Aggregation Server.

Client:
: The entity that provides user data to the system.

Measurement:
: The unencrypted, potentially-sensitive data that the client is asked to report.

Report:
: The encrypted measurement being sent by the client.

Auxiliary Data:
: Arbitrary data that clients may send as part of their report, but which is only
  revealed when at least K encrypted measurements of the same value are received.

REPORT_THRESHOLD:
: The minimum number of reports that an Aggregation Server needs before revealing
  client data. This value is chosen by the application.

> # Cryptographic Dependencies

Reviewer 2: Do we need to use double quotations for all of the cryptographic material in this section. For example when compiling the following:
```
- `Seal(key, nonce, aad, pt)`: Encrypt and authenticate plaintext
  `pt` with associated data `aad` using symmetric key `key` and nonce
  `nonce`, yielding ciphertext and tag `ct`.
```
it looks like this in the .txt file:
```
   *  Seal(key, nonce, aad, pt): Encrypt and authenticate plaintext pt
      with associated data aad using symmetric key key and nonce nonce,
      yielding ciphertext and tag ct.
```
Which doesn't seem super clear?

Author: How is it not clear?

STAR depends on the following cryptographic protocols and primitives:

- Threshold secret sharing (TSS); {{deps-tss}}
- Oblivious Pseudorandom Function (OPRF); {{deps-oprf}}
- Key Derivation Function (KDF); {{deps-kdf}}
- Key-Committing Authenticated Encryption with Associated Data (KCAEAD); {{deps-aead}}

This section describes the syntax for these protocols and primitives in more detail.

> ## Threshold Secret Sharing {#deps-tss}

Reviewer 2: Is the idea here to specify a standard secret-sharing scheme first, and then a _robust_ scheme afterwards? I was wondering if we could just go straight for robustness, which will allow us to relax authenticity below if `k` shares are provided that are valid in a group of `k + t` for some small values of `t`.

Author: We should define whatever scheme we need in this section. FWIW, I'm not sure we definitely need a robust scheme for all applications, but I'll leave it to y'all to sort out.

  for more details.
- Noe: The size of a serialized OPRF group element output from SerializeElement.

This specification uses the base (non-verifiable) OPRF from {{OPRF, Section 3}} with the
OPRF(ristretto255, SHA-512) as defined in {{OPRF, Section 4.1.1}}.

## Key Derivation Function {#deps-kdf}

A Key Derivation Function (KDF) is a function that takes some source of initial
keying material and uses it to derive one or more cryptographically strong keys.
This specification uses a KDF with the following API and parameters:

- Extract(salt, ikm): Extract a pseudorandom key of fixed length `Nx` bytes from
  input keying material `ikm` and an optional byte string `salt`.
- Expand(prk, info, L): Expand a pseudorandom key `prk` using the optional string `info`
  into `L` bytes of output keying material.
- Nx: The output size of the `Extract()` function in bytes.

This specification uses HKDF-SHA256 {{!HKDF=RFC5869}} as the KDF function, where Nx = 32.

## Key-Committing Authenticated Encryption with Associated Data {#deps-aead}

A Key-Committing Authenticated Encryption with Associated Data (KCAEAD) scheme is an algorithm

> for encrypting and authenticating plaintext with some additional auxiliary data.

Reviewer 2: This seems like it is going to get confusing, because auxiliary data support is explicitly provided in STAR, but the auxiliary data here is referring to additional data that can be included separately with the AEAD encryption process. It looks like you are suggesting that the auxiliary data is appended using the AEAD additional data functionality, rather than within the encrypted ciphertext as it is in the STAR paper, is that correct?

Author: No, thats not the intent. This is just defining the syntax of an AEAD, which explicitly includes additional data.

### Randomness Protocol

This procedure works as follows. Let `msg` be the client's measurement to be used for deriving
the randomness `rand`.

Clients first generate the a context for invoking the OPRF protocol as follows:

```
client_context = SetupOPRFClient(0x0001) // OPRF(ristretto255, SHA-512) ciphersuite
```

Clients then blind their measurement using this context as follows:

```
(blinded, blinded_element) = client_context.Blind(msg)
```

Clients then compute `randomness_request = SerializeElement(blinded_element)` and send it
to the Randomness Server URI in a HTTP POST message using content type "message/star-randomness-request".
An example request is shown below.

~~~
:method = POST
:scheme = https
:authority = randomness.example
:path = /
accept = message/star-randomness-response
content-type = message/star-randomness-response
content-length = Noe

<Bytes containing a serialized blinded element>
~~~

Upon receipt, the Randomness Server evaluates and returns a response.
It does so by first creating a context for running the ORPF protocol as follows:

```
server_context = SetupOPRFServer(0x0001, sk, pk) // OPRF(ristretto255, SHA-512) ciphersuite
```

> Here, `sk` and `pk` are private and public keys generated as described in {{randomness-configuration}}.

Author: Fixed.

### Reporting Protocol

This reporting protocol works as follows. First, the client stretches `rand` into three values
`key_seed`, `share_coins`, and `tag`, and additionally derives an KCAEAD key and nonce
from `key_seed`.

```
// Randomness derivation
rand_prk = Extract(nil, rand)
key_seed = Expand(rand_prk, "key_seed", 16)
share_coins = Expand(rand_prk, "share_coins", 16)
tag = Expand(rand_prk, "tag", 16)

// Symmetric encryption key derivation
key_prk = Extract(nil, key_seed)
key = Expand(key_prk, "key", Nk)
nonce = Expand(key_prk, "nonce", Nn)
```

The client then generates a secret share of `key_seed` using `share_coins` as randomness as follows:

```
rand_share = Share(REPORT_THRESHOLD, TBD, key_seed, share_coins, nil)
```

[[OPEN ISSUE: what should N be for the TSS scheme?]]

The client then encrypts `msg` and `aux` using the KCAEAD key and nonce as follows:

```
report_data = msg || aux
encrypted_report = Seal(key, nonce, nil, report_data)
```

Finally, the client constructs a report consisting of `encrypted_report`, `rand_share`,
and `tag`, and sends this to the Anonymizing Server in the subsequent epoch, i.e., after
the Randomness Server has rotated its OPRF key.

> [[OPEN ISSUE: how does the client know when this rotation happens?]]

Reviewer 2: This is one of the advantages of using a VOPRF, they can check that the correct key is being used for the time period.

Author: That would be a good thing to add.

Author: Let's do so in a followup PR.

  opaque rand_share[Nshare];
  opaque tag[16];
} Report;
```

Specifically, Clients send a Report to the Aggregation Server using an HTTP POST message
with content type "message/star-report". An example message is below.

~~~
:method = POST
:scheme = https
:authority = aggregator.example
:path = /
content-type = message/star-report
content-length = <Length of body>

<Bytes containing a Report>
~~~

This message is sent to the Aggregation Server through the Anonymizing Proxy. See {{proxy-options}}
for different types of proxy options.

## Aggregation Phase

Aggregation is the final phase of STAR. It happens offline and does not require any
communication between different STAR entities. It proceeds as follows. First, the
Aggregation Server groups reports together based on their `tag` value. Let `report_set`
denote a set of at least REPORT_THRESHOLD reports that have a matching `tag` value.

Given this set, the Aggregation Server begins by running the secret share recovery algoritm
as follows:

```
key_seed = Recover(report_set)
```

If this fails, the Aggregation Server chooses a new candidate report share set and
reruns the aggregation process.

> [[OPEN ISSUE: how does the server choose new candidate sets when share recovery fails?]]

Reviewer 2: I think robust secret-sharing solves this, e.g. https://www.cwi.nl/system/files/Fehr.pdf.

Otherwise, the Aggregation Server derives the same KCAEAD key and nonce from `key_seed` to
decrypt each of the report ciphertexts in `report_set`.

```
key_prk = Extract(nil, key_seed)
key = Expand(key_prk, "key", Nk)
nonce = Expand(key_prk, "nonce", Nn)
```

Each report ciphertext is decrypted as follows:

```
report_data = Seal(key, nonce, nil, ct)
msg || aux = report_data
```

If this fails for any report, the Aggregation Server chooses a new candidate report share set and
reruns the aggregation process. Otherwise, the Aggregation Server then outputs `msg` and each of
the `aux` values for the corresponding reports.

> [[OPEN ISSUE: what happens if the msg is different for any two reports in the set? This should not happen if using a KCAEAD, but good to handle nevertheless]]

Reviewer 2: Is this referring to if a client encrypts a different plaintext than the one used to create the share? If so, how would a KCAEAD help?

Author: A KCAEAD ensures that a given ciphertext can't decrypt to two different plaintexts under two different keys. So if the attacker can somehow force the server to use either of these keys, the server can't figure out which plaintext was correct. (This is somewhat tied to the secret sharing scheme.)
